source

```
@@ -65862,37 +65869,256 @@ console.log(pixels.data[2]);
+  run these steps:</p>
+
+  <ol>
+   <li><p>Let <var>xmlFilters</var> be an empty list.</p></li>
+
+   <li><p>If <var>filters</var> is a <code>CanvasFilterInput</code>, then set filters to « filters
+   ».</p></li>
+
+   <li>
+    <p>For each <var>filterDict</var> of <var>filters:</var></p>
+
+    <ol>
+     <li><p>Let <var>name</var> be the value of <var>filterDict</var>["name"].</p></li>
+
+     <li><p>If <var>name</var> is not one of <span>supported filter names</span>, then
+     continue.</p></li>
+
+     <li><p>Let <var>xmlName</var> be the concatenation of "fe", the first code unit of
+     <var>name</var> converted to ASCII uppercase, and all code units of name after the first
+     one.</p></li>
+
+     <li><p>Let <var>xmlFilter</var> be a new <span>XML element data for filters</span> whose <span
+     data-x="xml-filter-name">name</span> is <var>xmlName</var> and whose <span
+     data-x="xml-filter-attributes">attributes</span> is an empty ordered map.</p></li>
+
+     <li><p>Append <var>xmlFilter</var> to <var>xmlFilters</var>.</p></li>
+
+     <li>
+      <p>For each <var>attrName</var> → <var>attrValue</var> of <var>filterDict</var>.</p>
+
+      <ol>
+       <li>
+        <p>If any of the following is true:</p>
+
+        <ul>
+         <li><p><var>attrName</var> is not listed as an attribute for the filter primitive given
+         by xmlName</p></li>
+
+         <li><p><var>attrName</var> is a core attribute</p></li>
+
+         <li><p><var>attrName</var> is a presentation attribute</p></li>
+
+         <li><p><var>attrName</var> is a filter primitive attribute</p></li>
+
+         <li><p><var>attrName</var> is "class", "style", "in", or "name"</p></li>
+        </ul>
+
+        <p>then continue.</p>
+       </li>
+
+       <li>
+        <p>If <var>attrName</var> is one of « "funcR", "funcG", "funcB", "funcA" »</p>
+
+        <ol>
+         <li><p>If <var>attrValue</var> is not of type record, continue.</p></li>
```

Hmm, so attrValue is of type any here. We can't type test it because converting from any to record is an operation that might trigger side effects.

I think the most natural thing to do here would be to set attrValue to attrValue converted (using concept-idl-convert) to `record<DOMString, DOMString>`. This would throw an exception instead of ignoring it; if you think it's important to ignore such cases, then you could catch the exception and continue.

---

An exception is good here. The general design pattern her with exceptions is to pass silently for the general (using a filter effect that isn't there) and throw for the specific. This is definitely specific.

Shouldn't it be `record<DOMString, any>`?

---

I think either would work? In particular I assume you aren't going to nest any further so you would only expect strings at the next level.